File: packages/cloud-adapters/observability/baremetal_test.go
package observability

import (
	"context"
	"encoding/json"
	"io"
	"net/http"
	"net/http/httptest"
	"sync"
	"testing"
)

// fakeOTLP captures every request that hits its handler — one per
// OTLP signal endpoint (`/otlp/v1/logs|metrics|traces`).
type fakeOTLP struct {
	mu       sync.Mutex
	requests []capturedRequest
}

type capturedRequest struct {
	path        string
	method      string
	contentType string
	body        map[string]any
}

func (f *fakeOTLP) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	f.mu.Lock()
	defer f.mu.Unlock()
	buf, _ := io.ReadAll(r.Body)
	var body map[string]any
	_ = json.Unmarshal(buf, &body)
	f.requests = append(f.requests, capturedRequest{
		path:        r.URL.Path,
		method:      r.Method,
		contentType: r.Header.Get("Content-Type"),
		body:        body,
	})
	w.WriteHeader(http.StatusOK)
}

func newFakeOTLP(t *testing.T) (*BaremetalAdapter, *fakeOTLP, func()) {
	t.Helper()
	fk := &fakeOTLP{}
	server := httptest.NewServer(fk)
	adapter := NewBaremetalAdapter(Config{
		Client:      server.Client(),
		LogsURL:     server.URL + "/otlp/v1/logs",
		MetricsURL:  server.URL + "/otlp/v1/metrics",
		TracesURL:   server.URL + "/otlp/v1/traces",
		ServiceName: "monok8s-test",
	})
	return adapter, fk, server.Close
}

func TestBaremetalAdapter_PushLogs(t *testing.T) {
	adapter, fk, cleanup := newFakeOTLP(t)
	defer cleanup()
	err := adapter.PushLogs(context.Background(), []LogRecord{
		{Timestamp: 1700000000000000000, Severity: "INFO", Body: "hello", Labels: map[string]string{"tenant_id": "acme"}},
	})
	if err != nil {
		t.Fatalf("PushLogs: %v", err)
	}
	if len(fk.requests) != 1 {
		t.Fatalf("requests: got %d want 1", len(fk.requests))
	}
	req := fk.requests[0]
	if req.path != "/otlp/v1/logs" || req.method != "POST" {
		t.Errorf("logs request: got %s %s", req.method, req.path)
	}
	if req.contentType != "application/json" {
		t.Errorf("content-type: got %q want application/json", req.contentType)
	}
	// Drill into the OTLP envelope and assert the log record made it through.
	resourceLogs := req.body["resourceLogs"].([]any)
	scopeLogs := resourceLogs[0].(map[string]any)["scopeLogs"].([]any)
	logRecords := scopeLogs[0].(map[string]any)["logRecords"].([]any)
	logRec := logRecords[0].(map[string]any)
	if logRec["timeUnixNano"] != "1700000000000000000" {
		t.Errorf("timeUnixNano: got %v", logRec["timeUnixNano"])
	}
	if logRec["severityText"] != "INFO" {
		t.Errorf("severityText: got %v", logRec["severityText"])
	}
	body := logRec["body"].(map[string]any)
	if body["stringValue"] != "hello" {
		t.Errorf("body.stringValue: got %v", body["stringValue"])
	}
}

func TestBaremetalAdapter_PushMetrics(t *testing.T) {
	adapter, fk, cleanup := newFakeOTLP(t)
	defer cleanup()
	err := adapter.PushMetrics(context.Background(), []MetricPoint{
		{Name: "tenant.users.active", Value: 42.0, Timestamp: 1700000000000000000, Labels: map[string]string{"tenant_id": "acme"}},
	})
	if err != nil {
		t.Fatalf("PushMetrics: %v", err)
	}
	if fk.requests[0].path != "/otlp/v1/metrics" {
		t.Errorf("metrics request path: got %s", fk.requests[0].path)
	}
	metrics := fk.requests[0].body["resourceMetrics"].([]any)[0].(map[string]any)["scopeMetrics"].([]any)[0].(map[string]any)["metrics"].([]any)
	m := metrics[0].(map[string]any)
	if m["name"] != "tenant.users.active" {
		t.Errorf("metric name: got %v", m["name"])
	}
	dp := m["gauge"].(map[string]any)["dataPoints"].([]any)[0].(map[string]any)
	if dp["asDouble"] != 42.0 {
		t.Errorf("asDouble: got %v", dp["asDouble"])
	}
}

func TestBaremetalAdapter_PushTraces(t *testing.T) {
	adapter, fk, cleanup := newFakeOTLP(t)
	defer cleanup()
	err := adapter.PushTraces(context.Background(), []SpanRecord{{
		TraceID:     "abc",
		SpanID:      "def",
		Name:        "create-tenant",
		StartTimeNs: 1700000000000000000,
		EndTimeNs:   1700000001000000000,
		Attributes:  map[string]string{"tenant_id": "acme"},
	}})
	if err != nil {
		t.Fatalf("PushTraces: %v", err)
	}
	if fk.requests[0].path != "/otlp/v1/traces" {
		t.Errorf("traces request path: got %s", fk.requests[0].path)
	}
	spans := fk.requests[0].body["resourceSpans"].([]any)[0].(map[string]any)["scopeSpans"].([]any)[0].(map[string]any)["spans"].([]any)
	s := spans[0].(map[string]any)
	if s["name"] != "create-tenant" || s["traceId"] != "abc" || s["spanId"] != "def" {
		t.Errorf("span fields: name=%v traceId=%v spanId=%v", s["name"], s["traceId"], s["spanId"])
	}
}

func TestBaremetalAdapter_EmptyBatchesAreNoOps(t *testing.T) {
	adapter, fk, cleanup := newFakeOTLP(t)
	defer cleanup()
	ctx := context.Background()
	if err := adapter.PushLogs(ctx, nil); err != nil {
		t.Errorf("PushLogs nil: %v", err)
	}
	if err := adapter.PushMetrics(ctx, nil); err != nil {
		t.Errorf("PushMetrics nil: %v", err)
	}
	if err := adapter.PushTraces(ctx, nil); err != nil {
		t.Errorf("PushTraces nil: %v", err)
	}
	if len(fk.requests) != 0 {
		t.Errorf("empty batches sent requests: %d", len(fk.requests))
	}
}

func TestBaremetalAdapter_UnconfiguredEndpointErrors(t *testing.T) {
	adapter := NewBaremetalAdapter(Config{
		ServiceName: "test",
		// All three URLs left blank.
	})
	if err := adapter.PushLogs(context.Background(), []LogRecord{{Body: "x"}}); err == nil {
		t.Error("PushLogs with empty URL: expected error, got nil")
	}
}

File: packages/cloud-adapters/observability/baremetal.test.ts
/**
 * @jest-environment node
 *
 * Runtime contract test for the bare-metal OTLP ObservabilityAdapter.
 * Hermetic via stubbed fetch — verifies each push method builds the
 * right OTLP/HTTP/JSON envelope and POSTs to the right signal endpoint.
 */
import { describe, beforeEach, it, expect } from "@jest/globals";

import { newBaremetalAdapter } from "./baremetal";

type FetchCall = { url: string; init: { method?: string; body?: string; headers?: Record<string, string> } };

interface FakeFetch {
  (url: string, init?: unknown): Promise<unknown>;
  calls: FetchCall[];
}

const newFakeFetch = (): FakeFetch => {
  const calls: FetchCall[] = [];
  const fn = ((url: string, init: { method?: string; body?: string; headers?: Record<string, string> } = {}) => {
    calls.push({ url, init });
    return Promise.resolve({ ok: true, status: 200, text: () => Promise.resolve("") });
  }) as FakeFetch;
  fn.calls = calls;
  return fn;
};

const baseCfg = {
  logsURL: "https://loki.test/otlp/v1/logs",
  metricsURL: "https://mimir.test/otlp/v1/metrics",
  tracesURL: "https://tempo.test/otlp/v1/traces",
  serviceName: "monok8s-test",
};

describe("cloud-adapters/observability/baremetal — runtime wiring contract", () => {
  let f: FakeFetch;

  beforeEach(() => {
    f = newFakeFetch();
  });

  it("pushLogs POSTs OTLP envelope to logsURL", async () => {
    const adapter = newBaremetalAdapter({ ...baseCfg, fetch: f as unknown as typeof fetch });
    await adapter.pushLogs([
      { timestamp: 1700000000000000000, severity: "INFO", body: "hello", labels: { tenant_id: "acme" } },
    ]);
    expect(f.calls).toHaveLength(1);
    expect(f.calls[0]!.url).toBe(baseCfg.logsURL);
    expect(f.calls[0]!.init.method).toBe("POST");
    expect(f.calls[0]!.init.headers!["Content-Type"]).toBe("application/json");

    const payload = JSON.parse(f.calls[0]!.init.body as string);
    const rec = payload.resourceLogs[0].scopeLogs[0].logRecords[0];
    expect(rec.timeUnixNano).toBe("1700000000000000000");
    expect(rec.severityText).toBe("INFO");
    expect(rec.body.stringValue).toBe("hello");
    expect(rec.attributes[0]).toEqual({ key: "tenant_id", value: { stringValue: "acme" } });
  });

  it("pushMetrics POSTs gauge dataPoints to metricsURL", async () => {
    const adapter = newBaremetalAdapter({ ...baseCfg, fetch: f as unknown as typeof fetch });
    await adapter.pushMetrics([
      { name: "tenant.users.active", value: 42, timestamp: 1700000000000000000, labels: { tenant_id: "acme" } },
    ]);
    expect(f.calls[0]!.url).toBe(baseCfg.metricsURL);
    const payload = JSON.parse(f.calls[0]!.init.body as string);
    const m = payload.resourceMetrics[0].scopeMetrics[0].metrics[0];
    expect(m.name).toBe("tenant.users.active");
    expect(m.gauge.dataPoints[0].asDouble).toBe(42);
  });

  it("pushTraces POSTs spans to tracesURL", async () => {
    const adapter = newBaremetalAdapter({ ...baseCfg, fetch: f as unknown as typeof fetch });
    await adapter.pushTraces([{
      traceId: "abc",
      spanId: "def",
      name: "create-tenant",
      startTimeNs: 1700000000000000000,
      endTimeNs: 1700000001000000000,
      attributes: { tenant_id: "acme" },
    }]);
    expect(f.calls[0]!.url).toBe(baseCfg.tracesURL);
    const span = JSON.parse(f.calls[0]!.init.body as string)
      .resourceSpans[0].scopeSpans[0].spans[0];
    expect(span.traceId).toBe("abc");
    expect(span.spanId).toBe("def");
    expect(span.name).toBe("create-tenant");
  });

  it("empty batches are no-ops (no fetch calls)", async () => {
    const adapter = newBaremetalAdapter({ ...baseCfg, fetch: f as unknown as typeof fetch });
    await adapter.pushLogs([]);
    await adapter.pushMetrics([]);
    await adapter.pushTraces([]);
    expect(f.calls).toHaveLength(0);
  });

  it("auth header propagates when configured", async () => {
    const adapter = newBaremetalAdapter({
      ...baseCfg,
      authHeader: "Basic dGVuYW50LWFjbWU6cGFzc3dvcmQ=",
      fetch: f as unknown as typeof fetch,
    });
    await adapter.pushLogs([{ timestamp: 1, severity: "INFO", body: "x", labels: {} }]);
    expect(f.calls[0]!.init.headers!["Authorization"]).toBe("Basic dGVuYW50LWFjbWU6cGFzc3dvcmQ=");
  });

  it("non-2xx responses surface as errors", async () => {
    f = ((url: string, init: { headers?: Record<string, string> } = {}) => {
      void url;
      void init;
      return Promise.resolve({ ok: false, status: 503, text: () => Promise.resolve("upstream down") });
    }) as unknown as FakeFetch;
    f.calls = [];
    const adapter = newBaremetalAdapter({ ...baseCfg, fetch: f as unknown as typeof fetch });
    await expect(
      adapter.pushLogs([{ timestamp: 1, severity: "INFO", body: "x", labels: {} }]),
    ).rejects.toThrow(/503/);
  });
});
